DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/memory_system/README.md

```diff
@@ -0,0 +1,256 @@
+In this code sample, we have two kernels `NaiveKernel` and `OptimizedKernel`.
+
+The sample design shows access to a buffer implemented in on-chip memory. Read and write operations relies on LSUs.
+
+This tutorial simplifies a real-world access pattern to the banked memory system. There is a `buffer` that is implemented in on-chip memory, which accepts read and write. We write to the buffer and read from the buffer in a loop, which we expect an initiation interval of 1. This means that the read and write to the memory system need to be pipelined. Therefore, a number of read and write to the memory system in the loop body should all happen in 1 clock cycle.
+
+Each memory block (in this case MLAB) has a read port and a write port. For true dual-port memory setup, we can issue one read and one write in a single clock cycle. As a result, when the number of access exceeds the capacity of a memory block, the memory will become the bottleneck. Consequently, efficient access to the memory system becomes a key to the system performance.
+
+### Naive Kernel
+
+The access pattern in the `NaiveKernel` is illustrated in the following diagram:
+
+![Naive Kernel Memory Access Pattern](./assets/access_pattern_naive.gif)
+
+In the `NaiveKernel`, the first array dimension (row) is unrolled in a loop. In this way, the kernel tries to access all five elements in one column of the buffer, namely `buffer[0][col]` to `buffer[4][col]`. As a result, there are 4 loads and 5 stores to the memory system in each loop iteration, and since the initiation interval is 1, they happens in a single clocl cycle. The memory layout for the `buffer` in this case has 8 banks. Looking at the Details from the Kernel Memory Viewer, we can learn that the bank bits are second to the fourth least significant bits $b_4$, $b_3$, $b_2$. This essentially means that integers in every other column will be in a different memory bank.
+
+|                | Address | Bank bits | Bank # |
+|----------------|---------|-----------|--------|
+| buffer[0][0]   | 0x0000  | 0b000     | 0      |
+| buffer[0][1]   | 0x0004  | 0b001     | 1      |
+| buffer[0][2]   | 0x0008  | 0b010     | 2      |
+| buffer[0][3]   | 0x000C  | 0b011     | 3      |
+| buffer[0][4]   | 0x0010  | 0b100     | 4      |
+| ...            |         |           |        |
+| buffer[0][499] | 0x07CC  | 0b011     | 3      |
+| buffer[1][0]   | 0x07D0  | 0b100     | 4      |
+| ...            |         |           |        |
+
+We can produce this table showing which integer is stored in which bank of the memory system. 
+
+![Naive Kernel Memory Layout](./assets/memory_system_naive.png)
+
+Observe that each access to one column of the `buffer` creates 2 stores and 3 loads to even numbered memory banks and 2 stores and 2 loads to odd numbered memory banks. These number of load and store surpasses the physical capacity of the memory blocks, and therefore, the compiler inferrs stallable LSUs as arbitration node in the Kernel Memory Viewer.
```

Comment:
```suggestion
```
you nailed me for this in a previous code sample: the descriptive text should come before the picture, so I know what I'm looking at 😊